hles/transformation/src/main/scala/org/broadinstitute/monster/dap/HLESurveyTransformationPipelineBuilder.scala

```diff
@@ -77,4 +86,34 @@ object HLESurveyTransformationPipelineBuilder extends PipelineBuilder[Args] {
           RawRecord(id.toLong, fields)
       }
   }
+
+  /** Read in records and group by study Id, with field name subgroups. */
+  def readEnvRecords(ctx: ScioContext, args: Args): SCollection[RawRecord] = {
+    val rawRecords: SCollection[Msg] = StorageIO
+      .readJsonLists(
+        ctx,
+        "Raw Records",
+        s"${args.inputPrefix}/records/*.json"
+      )
+
+    // Group by study ID (record number) and arm (address_year_month)
+    // to get the format: (studyId, arm_id, Iterable((fieldName, Iterable(value))))
+    // (study_id, arm_id, iterable())
+    rawRecords
+      .groupBy(foo => {
+        (foo.read[String]("record"), foo.read[String]("redcap_event_name"))
+      })
+      //.groupBy(read[String]("redcap_event_name"))
+      .map {
+        case ((id, eventName), rawRecordValues) =>
+          val fields: Map[String, Array[String]] = rawRecordValues
+            .groupBy(_.read[String]("field_name"))
+            .map {
+              case (fieldName, rawValues) =>
+                (fieldName, rawValues.map(_.read[String]("value")).toArray.sorted)
+            }
+          val bar = fields + ("redcap_event_name" -> Array(eventName))
```

Comment:
🔧 Same comment re: naming; you might do away with this intermediate val altogether and do the concat inline when you construct the `RawRecord` instance.